### PR TITLE
[Tunix] Add multi-modal (vision) support to Gemma 4.

### LINKED ISSUE #1543: MM SFT of Gemma4
**Describe the solution you'd like**
Seems like there is no support in SFT/LORA/RL of gemma4 model with multi-modal inputs (text+image for example). There is support for gemma3, I would appriciate if you can supply such support and pointers

**Checklist**

- [ V] I have searched the existing issues for similar feature requests.
- [V ] This is not a support question (please use the "bug template" for that).

### PR DESCRIPTION
**Pull request by msghik**
Resolves #1543

Enables SFT / LoRA / RL of Gemma 4 with text + image inputs, mirroring the SigLIP-based integration already present in Gemma 3. The training and sampling code paths (`peft_trainer`, `sampler`, `rl/common`) already forward an `images` kwarg to the model — they just had no Gemma 4 model that accepted it. This change closes that gap at the model level so they work end-to-end on Gemma 4.

**What changed**

`tunix/models/gemma4/model.py`
- `ModelConfig` gains an optional `vision_config: SigLIPConfig | None`, exposed via a `text_only: bool = True` parameter on the `gemma4_e2b` / `gemma4_e4b` / `gemma4_31b` / `gemma4_26b_a4b` factory classmethods (matches the existing `gemma3_*_pt/it(text_only=...)` pattern).
- `ShardingConfig` gains a `siglip` field for the vision encoder's sharding.
- `RMSNorm` now accepts either a `ShardingConfig` (existing call sites) or a sharding tuple (new vision call site) — fully backward compatible.
- `Embedder` optionally builds the `mm_input_projection` and `mm_soft_embedding_norm` layers and exposes `encode_vision`.
- `Gemma4.__init__` constructs a `SigLiP` encoder when `vision_config` is set; `Gemma4.__call__` accepts an `images` kwarg and merges the soft vision tokens into the text embeddings at the placeholder positions.
- Adds `get_attention_mask` (bidirectional over image spans, causal over text) and `get_model_input` (used for LoRA tracing) on the model.

`tunix/models/gemma4/params_safetensors.py`
- `_get_key_and_transform_mapping` adds vision-tower and multi-modal projector mappings when `vision_config` is set (no change for text-only configs).

`tests/models/gemma4/model_test.py`
- 4 new tests: multi-modal forward pass, text-only construction has no vision encoder, helpful error when images are passed to a text-only model, and the text-only attention-mask shape.

**Design notes**

- The SigLIP encoder (`vision.py`), the embedding-merge helper (`merge_embeddings.py`), and the bidirectional-causal attention mask (`utils.py`) are reused from `tunix.models.gemma3` to avoid ~900 lines of duplication; they have no Gemma 3-specific assumptions. Happy to refactor these into a shared `tunix.models.common` namespace if reviewers prefer.
- The change is strictly additive: `vision_config` defaults to `None`, `images` defaults to `None`. Existing Gemma 4 call sites and tests are unaffected.

**Usage**

```python
from tunix.models.gemma4 import model as model_lib
from tunix.models.gemma4 import params_safetensors

config = model_lib.ModelConfig.gemma4_e4b(text_only=False)  # enables SigLIP
model = params_safetensors.create_model_from_safe_tensors(
    file_dir=ckpt_path, config=config, mesh=mesh, dtype=jnp.bfloat16,
)

logits, _ = model(
    tokens,
    positions=positions,
    attention_mask=model.get_attention_mask(tokens),
    images=images,
)

### CODE REVIEW

**Comment by tianshub** [comment]
Hi @msghik  thanks for adding the vision support. Can you paste a sample output for the multi-modal change?

**Review by tianshub — COMMENTED** [review]
Hi @msghik thanks for adding the vision support. Can you paste a sample output for the multi-modal change?

**Comment by msghik** [comment]
Hi @tianshub, sorry for the late response. I had been busy past couple of weeks. I figured out that there is a bug in this pull request that makes inference slow with my A100 GPU, something like 3 tokens/sec. I resolved the issue and open another pull request [Link](https://github.com/google/tunix/pull/1608), I believe we should contniue our conversation there. 

regarding your questoin: 

Here's an end-to-end sample from the multimodal path — google/gemma-4-E2B-it captioning a real photo (the standard COCO image,
 [image](http://images.cocodataset.org/val2017/000000039769.jpg): two cats sleeping on a pink couch with two TV remotes):

```
 $ python examples/gemma4/multimodal_generate.py \
  --ckpt ~/gemma-4-E2B-it --image cats.jpg \
   --prompt "What is in this image?"
```

 Detected Gemma4 text variant: E2B
 Image: 266 soft tokens (2394 raw patches)
Prompt: 'What is in this image?'  (prompt length: 273 tokens)

 === Caption ===
 This image contains two tabby cats lying on a pink surface. There are also
 some remote controls visible in the background.

 The image features, text prompt, and KV cache all flow through correctly — it picks up both cats, the pink surface, and the remotes. With a more detailed prompt it
 gives a full paragraph (per-cat coat patterns, positions, etc.).

I've also run the same code path on the large google/gemma-4-31B-it (60 layers) with 4-way tensor parallelism across 4×A100, so it works end-to-end across the Gemma4
size range.

**Comment by tianshub** [comment]
Hi @msghik thanks for the update. We actually have added multi-modal support (vision + audio) for gemma4 previously, and you can find example notebooks here: https://github.com/google/tunix/blob/main/tunix/models/gemma4/sampling_example.ipynb

please feel free to give a try and if there's any issue please report back.

**Comment by msghik** [comment]
Hi @tianshub, I see!

This is the third time that I try to contribute to Tunix and start an issue way earlier, however, I receive a message like this we already implemented it and didn't see your merge request. 

Could we please talk about something that I could work on that no one else would, so I could get my contribution to Tunix? it's a little bit frustrating from my side. 

Tnx

**Comment by tianshub** [comment]
Hi @msghik , sorry to hear that you ran into duplicated work issues. Since we have internal customers and sometimes a tight timeline, we may choose to duplicate the development with OSS contributors. I truly appreciate your effort and desire to contribute towards Tunix. Right now I don't have an active feature list, but if you have some ideas for post-training that you'd like to explore, please submit a RFC and we can collaborate there. Thanks!
